File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.java
package org.wikipedia.search;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.cursoradapter.widget.CursorAdapter;
import androidx.fragment.app.Fragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.CursorLoader;
import androidx.loader.content.Loader;

import org.wikipedia.R;
import org.wikipedia.WikipediaApp;
import org.wikipedia.database.contract.SearchHistoryContract;
import org.wikipedia.databinding.FragmentSearchRecentBinding;
import org.wikipedia.util.FeedbackUtil;

import io.reactivex.Completable;
import io.reactivex.schedulers.Schedulers;

import static org.wikipedia.Constants.RECENT_SEARCHES_FRAGMENT_LOADER_ID;

/** Displays a list of recent searches */
public class RecentSearchesFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor> {
    public interface Callback {
        void switchToSearch(@NonNull String text);

        void onAddLanguageClicked();
    }

    private Callback callback;
    private RecentSearchesAdapter adapter;

    private FragmentSearchRecentBinding binding;
    private ListView recentSearchesList;
    private View searchEmptyView;
    private View recentSearchesContainer;
    private View recentSearches;
    private TextView addLanguagesButton;
    private TextView emptyViewMessage;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentSearchRecentBinding.inflate(inflater, container, false);

        recentSearchesList = binding.recentSearchesList;
        searchEmptyView = binding.searchEmptyContainer;
        recentSearchesContainer = binding.recentSearchesContainer;
        recentSearches = binding.recentSearches;
        final ImageView deleteButton = binding.recentSearchesDeleteButton;
        addLanguagesButton = binding.addLanguagesButton;
        emptyViewMessage = binding.searchEmptyMessage;

        addLanguagesButton.setOnClickListener(v -> {
            if (callback != null) {
                callback.onAddLanguageClicked();
            }
        });

        deleteButton.setOnClickListener((view) ->
                new AlertDialog.Builder(requireContext())
                        .setMessage(getString(R.string.clear_recent_searches_confirm))
                        .setPositiveButton(getString(R.string.clear_recent_searches_confirm_yes), (dialog, id) ->
                                Completable.fromAction(() -> WikipediaApp.getInstance().getDatabaseClient(RecentSearch.class).deleteAll()).subscribeOn(Schedulers.io()).subscribe())
                        .setNegativeButton(getString(R.string.clear_recent_searches_confirm_no), null)
                        .create().show());
        FeedbackUtil.setToolbarButtonLongPressToast(deleteButton);

        return binding.getRoot();
    }

    public void show() {
        recentSearchesContainer.setVisibility(View.VISIBLE);
    }

    public void hide() {
        recentSearchesContainer.setVisibility(View.GONE);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        adapter = new RecentSearchesAdapter(getContext(), null, true);
        recentSearchesList.setAdapter(adapter);

        recentSearchesList.setOnItemClickListener((parent, view, position, id) -> {
            RecentSearch entry = (RecentSearch) view.getTag();
            if (callback != null) {
                callback.switchToSearch(entry.getText());
            }
        });

        LoaderManager supportLoaderManager = LoaderManager.getInstance(this);
        supportLoaderManager.initLoader(RECENT_SEARCHES_FRAGMENT_LOADER_ID, null, this);
        supportLoaderManager.restartLoader(RECENT_SEARCHES_FRAGMENT_LOADER_ID, null, this);
    }

    @Override
    public void onDestroyView() {
        LoaderManager.getInstance(this).destroyLoader(RECENT_SEARCHES_FRAGMENT_LOADER_ID);
        binding = null;
        super.onDestroyView();
    }

    @NonNull
    @Override
    public Loader<Cursor> onCreateLoader(int i, Bundle bundle) {
        Uri uri = SearchHistoryContract.Query.URI;
        String order = SearchHistoryContract.Query.ORDER_MRU;
        return new CursorLoader(requireContext(), uri, null, null, null, order);
    }

    @Override
    public void onLoadFinished(@NonNull Loader<Cursor> cursorLoaderLoader, Cursor cursorLoader) {
        if (!isAdded()) {
            return;
        }
        adapter.swapCursor(cursorLoader);
        boolean searchesEmpty = recentSearchesList.getCount() == 0;
        searchEmptyView.setVisibility(searchesEmpty ? View.VISIBLE : View.INVISIBLE);
        updateSearchEmptyView(searchesEmpty);
        recentSearches.setVisibility(!searchesEmpty ? View.VISIBLE : View.INVISIBLE);
    }

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    @SuppressWarnings("checkstyle:magicnumber")
    private void updateSearchEmptyView(boolean searchesEmpty) {
        if (searchesEmpty) {
            searchEmptyView.setVisibility(View.VISIBLE);
            if (WikipediaApp.getInstance().language().getAppLanguageCodes().size() == 1) {
                addLanguagesButton.setVisibility(View.VISIBLE);
                emptyViewMessage.setText(getString(R.string.search_empty_message_multilingual_upgrade));
            } else {
                addLanguagesButton.setVisibility(View.GONE);
                emptyViewMessage.setText(getString(R.string.search_empty_message));
            }
        } else {
            searchEmptyView.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public void onLoaderReset(@NonNull Loader<Cursor> cursorLoaderLoader) {
        adapter.changeCursor(null);
    }

    void updateList() {
        adapter.notifyDataSetChanged();
    }

    private class RecentSearchesAdapter extends CursorAdapter {
        RecentSearchesAdapter(Context context, Cursor c, boolean autoRequery) {
            super(context, c, autoRequery);
        }

        @Override
        public View newView(Context context, Cursor cursor, ViewGroup viewGroup) {
            return LayoutInflater.from(getActivity()).inflate(R.layout.item_search_recent, viewGroup, false);
        }

        @Override
        public void bindView(View view, Context context, Cursor cursor) {
            TextView textView = (TextView) view;
            RecentSearch entry = getEntry(cursor);
            textView.setText(entry.getText());
            view.setTag(entry);
        }

        @Override
        public CharSequence convertToString(Cursor cursor) {
            return getEntry(cursor).getText();
        }

        RecentSearch getEntry(Cursor cursor) {
            return RecentSearch.DATABASE_TABLE.fromCursor(cursor);
        }
    }
}

File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt
package org.wikipedia.search

import android.graphics.Typeface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.ConcatAdapter
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.launch
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.analytics.eventplatform.RabbitHolesEvent
import org.wikipedia.database.AppDatabase
import org.wikipedia.databinding.FragmentSearchRecentBinding
import org.wikipedia.dataclient.ServiceFactory
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.dataclient.mwapi.MwQueryResult
import org.wikipedia.page.Namespace
import org.wikipedia.search.db.RecentSearch
import org.wikipedia.util.FeedbackUtil.setButtonTooltip
import org.wikipedia.util.ResourceUtil
import org.wikipedia.util.log.L
import org.wikipedia.views.SwipeableItemTouchHelperCallback
import java.util.concurrent.ConcurrentHashMap

class RecentSearchesFragment : Fragment() {
    interface Callback {
        fun switchToSearch(text: String)
        fun onAddLanguageClicked()
        fun getLangCode(): String
    }

    private var _binding: FragmentSearchRecentBinding? = null
    val binding get() = _binding!!
    private val namespaceHints = listOf(Namespace.USER, Namespace.PORTAL, Namespace.HELP)
    private val namespaceMap = ConcurrentHashMap<String, Map<Namespace, String>>()
    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable -> L.e(throwable) }
    var callback: Callback? = null
    val recentSearchList = mutableListOf<RecentSearch>()
    private var suggestedSearchTerm: String? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentSearchRecentBinding.inflate(inflater, container, false)

        binding.recentSearchesDeleteButton.setOnClickListener {
            MaterialAlertDialogBuilder(requireContext())
                    .setMessage(getString(R.string.clear_recent_searches_confirm))
                    .setPositiveButton(getString(R.string.clear_recent_searches_confirm_yes)) { _, _ ->
                        lifecycleScope.launch(coroutineExceptionHandler) {
                            AppDatabase.instance.recentSearchDao().deleteAll()
                            updateList()
                        }
                    }
                    .setNegativeButton(getString(R.string.clear_recent_searches_confirm_no), null)
                    .show()
        }
        binding.addLanguagesButton.setOnClickListener { onAddLangButtonClick() }
        binding.recentSearchesRecycler.layoutManager = LinearLayoutManager(requireActivity())
        binding.namespacesRecycler.layoutManager = LinearLayoutManager(requireActivity(), RecyclerView.HORIZONTAL, false)

        suggestedSearchTerm = requireActivity().intent.getStringExtra(SearchActivity.EXTRA_SUGGESTED_QUERY)
        binding.recentSearchesRecycler.adapter = if (suggestedSearchTerm.isNullOrEmpty()) RecentSearchAdapter() else ConcatAdapter(SuggestedSearchAdapter(), RecentSearchAdapter())
        binding.recentSearchesHeaderContainer.isVisible = suggestedSearchTerm.isNullOrEmpty()

        val touchCallback = SwipeableItemTouchHelperCallback(requireContext())
        touchCallback.swipeableEnabled = true
        val itemTouchHelper = ItemTouchHelper(touchCallback)
        itemTouchHelper.attachToRecyclerView(binding.recentSearchesRecycler)
        setButtonTooltip(binding.recentSearchesDeleteButton)
        return binding.root
    }

    fun show() {
        binding.recentSearchesContainer.visibility = View.VISIBLE
    }

    fun hide() {
        binding.recentSearchesContainer.visibility = View.GONE
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun updateSearchEmptyView(searchesEmpty: Boolean) {
        if (searchesEmpty) {
            binding.searchEmptyContainer.visibility = View.VISIBLE
            if (WikipediaApp.instance.languageState.appLanguageCodes.size == 1) {
                binding.addLanguagesButton.visibility = View.VISIBLE
                binding.searchEmptyMessage.text = getString(R.string.search_empty_message_multilingual_upgrade)
            } else {
                binding.addLanguagesButton.visibility = View.GONE
                binding.searchEmptyMessage.text = getString(R.string.search_empty_message)
            }
        } else {
            binding.searchEmptyContainer.visibility = View.INVISIBLE
        }
    }

    private fun onAddLangButtonClick() {
        callback?.onAddLanguageClicked()
    }

    fun reloadRecentSearches() {
        lifecycleScope.launch(coroutineExceptionHandler) {
            updateList()
        }
    }

    suspend fun updateList() {
        val nsMap: Map<String, MwQueryResult.Namespace>
        val langCode = callback?.getLangCode().orEmpty()

        if (!namespaceMap.containsKey(langCode)) {
            val map = mutableMapOf<Namespace, String>()
            namespaceMap[langCode] = map
            nsMap = ServiceFactory.get(WikiSite.forLanguageCode(langCode)).getPageNamespaceWithSiteInfo(null).query?.namespaces.orEmpty()
            namespaceHints.forEach {
                map[it] = nsMap[it.code().toString()]?.name.orEmpty()
            }
        }

        val searches: List<RecentSearch> = AppDatabase.instance.recentSearchDao().getRecentSearches()

        recentSearchList.clear()
        recentSearchList.addAll(searches)

        val searchesEmpty = recentSearchList.isEmpty() && suggestedSearchTerm.isNullOrEmpty()
        binding.namespacesRecycler.adapter = NamespaceAdapter()
        if (binding.recentSearchesRecycler.adapter is ConcatAdapter) {
            (binding.recentSearchesRecycler.adapter as ConcatAdapter).adapters.forEach {
                it.notifyDataSetChanged()
            }
        } else {
            binding.recentSearchesRecycler.adapter?.notifyDataSetChanged()
        }
        binding.searchEmptyContainer.isInvisible = !searchesEmpty
        updateSearchEmptyView(searchesEmpty)
        binding.recentSearches.isInvisible = searchesEmpty
    }

    private open inner class RecentSearchItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, SwipeableItemTouchHelperCallback.Callback {
        private lateinit var recentSearch: RecentSearch

        open fun bindItem(position: Int) {
            recentSearch = recentSearchList[position]
            itemView.setOnClickListener(this)
            (itemView as TextView).text = recentSearch.text
        }

        override fun onClick(v: View) {
            RabbitHolesEvent.submit("recent_search_click", "search",
                source = if (this is SuggestedSearchItemViewHolder) "suggested" else "standard",
                recShown = !suggestedSearchTerm.isNullOrEmpty())

            callback?.switchToSearch((v as TextView).text.toString())
        }

        override fun onSwipe() {
            lifecycleScope.launch(coroutineExceptionHandler) {
                AppDatabase.instance.recentSearchDao().delete(recentSearch)
                updateList()
            }
        }

        override fun isSwipeable(): Boolean { return true }
    }

    private inner class SuggestedSearchItemViewHolder(itemView: View) : RecentSearchItemViewHolder(itemView) {
        override fun bindItem(position: Int) {
            (itemView as TextView).apply {
                setOnClickListener(this@SuggestedSearchItemViewHolder)
                setBackgroundColor(ResourceUtil.getThemedColor(requireContext(), R.attr.addition_color))
                setTypeface(null, Typeface.NORMAL)
                text = suggestedSearchTerm
            }
        }

        override fun isSwipeable(): Boolean { return false }
    }

    private inner class SuggestedSearchHeadingViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bindItem(position: Int) {
            (itemView as TextView).apply {
                setBackgroundColor(ResourceUtil.getThemedColor(requireContext(), R.attr.paper_color))
                setTypeface(null, Typeface.BOLD)
                text = if (position == 0) getString(R.string.recent_searches_related_to_reading) else getString(R.string.recent_searches_title)
            }
        }
    }

    private inner class SuggestedSearchAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
        override fun getItemCount(): Int {
            // 1) "Suggested search" heading, 2) Suggested query, 3) "Recent searches" heading
            return if (recentSearchList.isEmpty()) 2 else 3
        }

        override fun getItemViewType(position: Int): Int {
            return if (position == 1) LIST_TYPE_ITEM else LIST_TYPE_HEADING
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            return if (viewType == LIST_TYPE_HEADING) {
                SuggestedSearchHeadingViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_search_recent, parent, false))
            } else {
                SuggestedSearchItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_search_recent, parent, false)
                )
            }
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, pos: Int) {
            if (holder is SuggestedSearchItemViewHolder) {
                holder.bindItem(pos)
            } else if (holder is SuggestedSearchHeadingViewHolder) {
                holder.bindItem(pos)
            }
        }
    }

    private inner class RecentSearchAdapter : RecyclerView.Adapter<RecentSearchItemViewHolder>() {
        override fun getItemCount(): Int {
            return recentSearchList.size
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecentSearchItemViewHolder {
            return RecentSearchItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_search_recent, parent, false))
        }

        override fun onBindViewHolder(holder: RecentSearchItemViewHolder, pos: Int) {
            holder.bindItem(pos)
        }
    }

    private inner class NamespaceItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        fun bindItem(ns: Namespace?) {
            val isHeader = ns == null
            (itemView as TextView).apply {
                setOnClickListener(this@NamespaceItemViewHolder)
                text = if (isHeader) getString(R.string.search_namespaces) else namespaceMap[callback?.getLangCode()]?.get(ns).orEmpty() + ":"
                isEnabled = !isHeader
                setTextColor(ResourceUtil.getThemedColor(requireContext(), if (isHeader) R.attr.primary_color else R.attr.progressive_color))
            }
        }

        override fun onClick(v: View) {
            callback?.switchToSearch((v as TextView).text.toString())
        }
    }

    private inner class NamespaceAdapter : RecyclerView.Adapter<NamespaceItemViewHolder>() {
        override fun getItemCount(): Int {
            return namespaceHints.size + 1
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NamespaceItemViewHolder {
            return NamespaceItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_namespace, parent, false))
        }

        override fun onBindViewHolder(holder: NamespaceItemViewHolder, pos: Int) {
            holder.bindItem(namespaceHints.getOrNull(pos - 1))
        }
    }

    companion object {
        const val LIST_TYPE_HEADING = 0
        const val LIST_TYPE_ITEM = 1
    }
}
